CRYPTOGRAPHY.md

```diff
@@ -2,120 +2,526 @@
+The blobs are individually encoded so it is possible to seek to a specific portion of the packfile and decode a blob without requiring the read of previous bytes to have the correct decompression or encryption state.
+
+Since blobs are created from a variety of data structures that are not only related to file content but also to filesystem structure and software internal structures,
+and because the backup process parallelizes packfiles filling, filesystem discovery and file content chunking,
+two contiguous blobs in a packfile almost never represent data that is contiguous in the data source being backed up.
+A file content has blobs spread across multiple packfiles and two blobs from the same file content within a same packfile are almost always interlaced with unrelated blobs encoded from other data content.
+
+
+#### Index section
+The index section contains a sequence of fixed-length records that contain the blob type, the blob version, a HMAC of the original data checksum using the repository master key as secret, the offset of the blob in the data section and its length.
+
+The application uses regular checksums within snapshots so they can be compared to checksums on the live filesystem:
+a user can compute sha256 of /etc/passwd and match it with a checksum within the snapshot.
+However,
+whenever a checksum is used as a lookup key to a packfile,
+it is first converted to a HMAC using the master key so the original checksums are never referenced in the packfiles.
+
+The packfile index is individually encoded and not meant to be used as the repository state files provide a global index that aggregates all packfiles indexes.
+They are mainly used as a replica for disaster recovery should repository state end up corrupted:
+if really needed, we can locate the offset of the index thanks to the packfile footer and decode it to access its content and understand the structure of the packfile.
+
+The index has its HMAC computed and stored in the footer for fast integrity check without needing to read the entire packfile.
+
+
+#### Footer section
+The footer is a fixed-size structure which allows to locate the offset of the index, its size and its HMAC.
+
+It does not contain sensitive data but is encrypted as to not leak the index offset and make it harder to determine the number of blobs in a packfile:
+a packfile of ~20MB may contain only 2 x ~10MB blobs or may containt 20x ~1MB blobs, only the index provides this information.
+
+
+### State files
+ 
+State files are immutable structured files that provide an event log of changes happening to a repository.
+
+Every time a backup is done,
+a new state file is created that contains records describing which packfiles were created,
+which blob HMACs point to which locations (packfile, offset, length),
+which snapshots were deleted (to allow a maintenance job to find orphaned packfiles to be removed).
```

Comment:
maybe with s/immutable/append-only/ is more clear

Reply:
resolved